src/inline-assembly.md

```diff
@@ -0,0 +1,474 @@
+# Inline assembly
+
+Support for inline assembly is provided via the [`asm!`] and [`global_asm!`] macros.
+It can be used to embed handwritten assembly in the assembly output generated by the compiler.
+
+[`asm!`]: ../core/arch/macro.asm.html
+[`global_asm!`]: ../core/arch/macro.global_asm.html
+
+Support for inline assembly is stable on the following architectures:
+- x86 and x86-64
+- ARM
+- AArch64
+- RISC-V
+
+The compiler will emit an error if `asm!` is used on an unsupported target.
+
+## Example
+
+```rust
+use std::arch::asm;
+
+// Multiply x by 6 using shifts and adds
+let mut x: u64 = 4;
+unsafe {
+    asm!(
+        "mov {tmp}, {x}",
+        "shl {tmp}, 1",
+        "shl {x}, 2",
+        "add {x}, {tmp}",
+        x = inout(reg) x,
+        tmp = out(reg) _,
+    );
+}
+assert_eq!(x, 4 * 6);
+```
+
+## Syntax
+
+The following ABNF specifies the general syntax:
+
+```text
+dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
+reg_spec := <register class> / "<explicit register>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand
+clobber_abi := "clobber_abi(" <abi> *("," <abi>) [","] ")"
+option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn" / "nostack" / "att_syntax" / "raw"
+options := "options(" option *("," option) [","] ")"
+asm := "asm!(" format_string *("," format_string) *("," [ident "="] operand) *("," clobber_abi) *("," options) [","] ")"
+global_asm := "global_asm!(" format_string *("," format_string) *("," [ident "="] operand) *("," options) [","] ")"
```

Comment:
`format_string` isn't defined in this mini-grammar. Perhaps add something like this?

```
format_string := STRING_LITERAL / RAW_STRING_LITERAL
```